lib/crawler.js

```diff
@@ -1271,17 +1273,23 @@ Crawler.prototype.handleResponse = function(queueItem, response, timeCommenced)
             crawler.cache.setCacheData(queueItem, responseBuffer);
         }
 
-        // Is the item allowed by depth conditions ?
         if (crawler.depthAllowed(queueItem)) {
-            var responseBody =
-                crawler.decodeResponses ? crawler.decodeBuffer(responseBuffer, stateData.contentType) : responseBuffer;
-
-            crawler.emit("fetchcomplete", queueItem, responseBody, response);
-
-            // We only process the item if it's of a valid mimetype
-            // and only if the crawler is set to discover its own resources
-            if (crawler.mimeTypeSupported(contentType) && crawler.discoverResources) {
-                crawler.queueLinkedItems(responseBuffer, queueItem);
+            // No matter the value of `crawler.decompressResponses`, we still
+            // decompress the response if it's gzipped or deflated. This is
+            // because we always provide the discoverResources method with a
+            // decompressed buffer
+            if (/(gzip|deflate)/.test(stateData.headers["content-encoding"])) {
+                zlib.unzip(responseBuffer, function(error, decompressedBuffer) {
+                    if (error) {
+                        crawler.emit("gziperror", queueItem, error, responseBuffer);
+                        emitFetchComplete(responseBuffer);
+                    } else {
+                        var responseBody = crawler.decompressResponses ? decompressedBuffer : responseBuffer;
+                        emitFetchComplete(responseBody, decompressedBuffer);
```

Comment:
I just noticed that in the scenario where the body is successfully decoded, we're passing the decompressed response body twice. Is this something we need to do? (If we don't make it configurable, we can avoid that weird logic altogether!) If we decide to leave it in, let's comment it.